Documentation/design-docs/PinnedHeap.md

```diff
@@ -0,0 +1,160 @@
+
+<span style="color:blue">**_When to collect_**</span>
+
+LOH is only collected during gen2 GCs. Since the pinned heap is supposed to be small we can consider collecting it in gen1 GCs as well. Or have something more dynamic like based on its size relative to the rest of the heap.
+
+<span style="color:blue">**_Accounting_**</span>
+
+The pinned heap is part of the GC heap so when the heap size is requested pinned heap size should be included, eg, <span style="color:red">`GC.GetTotalMemory`</span> should include sizes of objects on the pinned heap. 
+
+For generational accounting the pinned heap should be considered its own generation as it has a different perf characteristic from the other generations. For example, it wouldn’t make sense for <span style="color:red">`generation_free_list_space`</span> for pinned heap to be one of the current generations. This way we can also avoid changing the current accounting, eg, <span style="color:red">`decide_on_compacting`</span>/<span style="color:red">`generation_to_condemn`</span> which cares about generational accounting.
+
+<span style="color:blue">**_Free list management_**</span>
+
+This is again very similar to what we do today with LOH allocations which uses a bucketed free list (<span style="color:red">`loh_alloc_list`</span>). We’d want different bucket sizes/probably different # of buckets. This also means 
+
+<span style="color:blue">**_BGC consideration_**</span>
+
+Since pinned heap will not have objects with references we will not need to go through <span style="color:red">`revisit_written_pages`</span> for it (which automatically happens as it only goes through gen2/3 segments today). We can let <span style="color:red">`background_sweep`</span> take care of sweeping on pinned heap by including its generation.
+
+<span style="color:blue">**_HardLimit consideration_**</span>
+
+Today HardLimit requires us to reserve all segments when the GC is initialized. And since we cannot predict how much SOH allocation vs LOH allocation we’d have we just reserve 1x for SOH and actually 2x for LOH (we use 1x for LOH when you use large pages). Obviously we do not want to reserve 1x for pinned heap. Since pinned heap should be very small compared to SOH, we could reserve a smaller segment size. Although the reserve size doesn’t matter that much and commit is automatically kept track of. We do need to think more carefully for the large pages case as we need to commit upfront.
+
+**2) BCL/VM changes.**
+
+**3) Framework component changes**
+
+On Desktop CLR we introduced a <span style="color:red">`PinnableBufferCache`</span> that was used by a few framework components which would be good candidates for investigation purpose. Since we are doing the work on coreclr we don’t have those changes but we can do a code review to see whether it’s seamless to use the new API.
```

Comment:
Majority of pinned buffers allocated by .NET Core come from array pool. Fixing ArrayPool to use this would be the most important part. We may consider to always allocate ArrayPool buffers that do not contain object references on pinned heap.

Reply:
thanks for the observations. we will definitely take these into account.